### PR TITLE
API layer changes and test cases for the new Subject sortBy column

### PR DESCRIPTION
**Pull request by Prashanth-Billa**
First version

### CODE REVIEW

**Comment by Prashanth-Billa** [comment]
Merge conflicts in pull request resolved

File: api/v1/swagger.yaml [hunk 1]
```diff
@@ -6998,6 +7043,11 @@ definitions:
         maxLength: 2082
         description: >
           Authentication success message.
+      token:
```
Comment (iamigo) [inline api/v1/swagger.yaml]: this shouldn't be here

File: migrations/20170203233255-addSortByToSubject.js [hunk 2]
```diff
@@ -10,8 +10,9 @@ module.exports = {
     return qi.sequelize.transaction(() => qi.addColumn('Subjects', 'sortBy', {
       type: Sequelize.STRING(constants.fieldlen.sortField),
       allowNull: true,
+      defaultValue: '',
```
Comment (iamigo) [inline migrations/20170203233255-addSortByToSubject.js]: you can't modify the migration--it's already been deployed
if you need more db changes, you have to create a new migration

File: tests/api/v1/subjects/put.js [hunk 3]
```diff
@@ -560,4 +600,4 @@ describe('api: PUT subjects with tags', () => {
       done();
     });
   });
-});
+});
```
Comment (iamigo) [inline tests/api/v1/subjects/put.js]: must have trailing newline

File: migrations/20170203233255-addSortByToSubject.js (hunk 2 above)
Comment (Prashanth-Billa) [inline migrations/20170203233255-addSortByToSubject.js]: resolved

File: tests/api/v1/subjects/put.js (hunk 3 above)
Comment (Prashanth-Billa) [inline tests/api/v1/subjects/put.js]: resolved

**Comment by Prashanth-Billa** [comment]
Style issues resolved for all test files. create, update in subject model. get, put, post, patch in API db subject

**Comment by iamigo** [comment]
@Prashanth-Billa test failures on Subject find?

File: migrations/20170213221219-newSortByColumnForSubject.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+      type: Sequelize.STRING(constants.fieldlen.sortField),
+      allowNull: true,
+      defaultValue: '',
+      validate: {
+        is: /^[0-9a-z_-]*$/i,
+      },
+    }));
+  },
+
+  down: function (qi, Sequelize) {
```
Comment (iamigo) [inline migrations/20170213221219-newSortByColumnForSubject.js]: This migration code is wrong--it should be modifying the table definition from its previous state. The column *already* exists (from your last migration). What specifically are you changing from what the previous commit defined?

File: migrations/20170213221219-newSortByColumnForSubject.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    /*
+      Add reverting commands here.
+      Return a promise to correctly handle asynchronicity.
+    */
+    return qi.sequelize.transaction(() => qi.removeColumn('Subjects', 'sortBy'));
```
Comment (iamigo) [inline migrations/20170213221219-newSortByColumnForSubject.js]: same thing here--should revert to whatever was there previously

File: tests/api/v1/subjects/get.js [hunk 6]
```diff
@@ -23,335 +23,381 @@ const ONE = 1;
...
-      vt.parentId = us.id;
-      return Subject.create(vt);
-    })
-    .then((createdVt) => {
-      vt.id = createdVt.id;
-      done();
-    })
-    .catch(done);
-  });
+    before((done) => {
```
Comment (iamigo) [inline tests/api/v1/subjects/get.js]: Tabs should be two spaces, never four.

File: tests/api/v1/subjects/get.js [hunk 7]
```diff
@@ -23,335 +23,381 @@ const ONE = 1;
...
-      .end((err/* , res */) => {
-        if (err) {
-          done(err);
-        }
-
-        done();
+    after(u.forceDelete);
+    after(tu.forceDeleteUser);
+
+    describe('duplicate tags fail', () => {
```
Comment (iamigo) [inline tests/api/v1/subjects/get.js]: Please make sure all tab indents meet style guidelines.

**Comment by iamigo** [comment]
Good job on code coverage!

File: migrations/20170213221219-newSortByColumnForSubject.js (hunk 4 above)
Comment (Prashanth-Billa) [inline migrations/20170213221219-newSortByColumnForSubject.js]: I hope the changeColumn method would work. The UP and DOWN states now reflect the new and previous state correspondingly. Pls review. Thanks

Comment (iamigo) [inline migrations/20170213221219-newSortByColumnForSubject.js]: You only have to include the attribute which have changed here: defaultValue and the validate regex.

**Comment by Prashanth-Billa** [comment]
ok, defaultValue and validate fields are only included.

File: api/v1/swagger.yaml [hunk 8]
```diff
@@ -4000,6 +4016,13 @@ paths:
                 description: >
                   Max 60 characters; case insensitive; allows alpha-numeric characters,
                   underscore (_) and dash (-).
+              sortBy:
```
Comment (pallavi2209) [inline api/v1/swagger.yaml]: Overall changes look good. I just felt that we can one more line in description on how this sortBy field is used. Something Like "Default empty. When set to some string, the subjects are sorted alphabetically on string value."
@Prashanth-Billa If you want, you can make this change in next PR, I will merge this now

Comment (Prashanth-Billa) [inline api/v1/swagger.yaml]: sure. sounds good! Will include the description change as part of story 3 - view layer. Thank you

Comment (iamigo) [inline api/v1/swagger.yaml]: Good suggestion. I don't think the core Refocus app would actually use this field to sort subjects in the admin console or even in api response... it's more about making it available as an option for the Lens developer to choose to use or not. Go back to the description in the user story and try to come up with some language to add to swagger based on that.

**Pull request merged**